### PR TITLE
ClipPath

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

**Comment by Hixie** [comment]
@abarth 

This is for customer: pink. Depends on https://github.com/flutter/engine/pull/2642.

cc @apwilson 


I'll add a test to make sure hit-testing works before I check this in.

File: packages/flutter/lib/src/material/app_bar.dart [hunk 1]
```diff
@@ -280,7 +280,30 @@ class AppBar extends StatelessWidget {
       child: appBar
     );
 
+    return new ClipPath(
+      clipper: new DemoPathClipper(),
```
Comment (abarth) [inline packages/flutter/lib/src/material/app_bar.dart]: Presumably you don't want to check this part in.

**Comment by eseidelGoogle** [comment]
Do we know if path-defined clippaths are fast in Ganesh?  I assume so, just that may affect what docs we leave the user in terms of performance expectations.

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 2]
```diff
@@ -994,6 +994,37 @@ class RenderClipOval extends _RenderCustomClip<Rect> {
...
+  bool hitTest(HitTestResult result, { Point position }) {
+    if (_cachedPath == null || !_cachedPath.contains(position))
+      return false;
+    return super.hitTest(result, position: position);
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (child != null) {
+      _cachedPath = _clip; // invokes the callback
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: Do we want to regenerate the path every time we paint?  Maybe we should have logic to clear out the cached path at some point, similar to custom paint.

**Comment by abarth** [comment]
> Do we know if path-defined clippaths are fast in Ganesh? I assume so, just that may affect what docs we leave the user in terms of performance expectations.

I would assume they're slow given that RRect clips are special cased.

**Comment by eseidelGoogle** [comment]
If that's the case, I would recommend we note that in the docs for this widget?

FYI @bsalomon

**Comment by bsalomon** [comment]
Arbitrary paths are not as fast as rects or rrects. However, if we can detect whether a path is really a rect or rrect.

File: packages/flutter/lib/src/rendering/proxy_box.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: Yeah I suppose it only needs to be done once per layout.

Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: I've updated all the clip render objects to cache the results of calling the delegate at layout time.

**Comment by Hixie** [comment]
I've added to the docs a warning that `ClipPath` is going to be slower than the alternatives, and provided links to the alternatives.

File: packages/flutter/lib/src/material/app_bar.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/app_bar.dart]: removed

**Comment by abarth** [comment]
LGTM

**Comment by Hixie** [comment]
Added a test also.

**Pull request merged**